<a href="https://colab.research.google.com/github/redinbluesky/handson-llm/blob/main/02_토큰과_임베딩.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#  목차
* [Chapter 2 서론](#chapter2)
* [Chapter 2-1 LLM 토큰화](#chapter2-1)
    * [Chapter 2-1-1 토크나이져가 언어 모델의 입력을 준비하는 방법](#chapter2-1-1)
    * [Chapter 2-1-2 LLM 다운로드하고 실행하기](#chapter2-1-2)
    * [Chapter 2-1-3 토크나이저가 텍스트를 분할하는 방법](#chapter2-1-3)
    * [Chapter 2-1-4 단어 토큰, 부분단어 토큰, 문자 토큰, 바이트 토큰](#chapter2-1-4)
    * [Chapter 2-1-5 훈련된 LLM 토크나이져 비교하기](#chapter2-1-5)
    * [Chapter 2-1-6 토크나이저 속성](#chapter2-1-6)
* [Chapter 2-2 토큰 임베딩](#chapter2-2)    
    * [Chapter 2-2-1 토크나이저의 어휘사전에 대한 임베딩을 내장한 언어 모델](#chapter2-2-1)    
    * [Chapter 2-2-2 언어 모델로 문맥을 고려한 단어 임베딩 만들기](#chapter2-2-2)
* [Chapter 2-3 텍스트 임베딩](#chapter2-3)       
* [Chapter 2-4 LLM을 넘어 활용되는 단어 임베딩](#chapter2-4)  
    * [Chapter 2-4-1 사전 훈련된 단어 임베딩 다운로드](#chapter2-4-1)      
    * [Chapter 2-4-2 word2vec 알고리즘과 대조 훈련](#chapter2-4-2)   
* [Chapter 2-5 추천시스템을 위한 시스템 임베딩](#chapter2-5)
    * [Chapter 2-5-1 임베딩으로 노래 추천하기](#chapter2-5-1)        
    * [Chapter 2-5-2 노래 임베딩 모델 훈련하기](#chapter2-5-2)   


## Chapter 2 서론 <a class="anchor" id="chapter2"></a>
1. 토큰과 임베딩은 대규모 언어 모델을 사용하느 데 중심이 되는 두가지 개념이다.
    - 토큰과 임베딩을 이해하지 못하면 대규모 언어 모델이 어떻게 작동하는지 이해하기 어렵다.
    - 언어 모델은 텍스트를 토큰이라는 작은 단위로 나우어 처리한다.
    - 언어 모델은 언어에 대한 계산을 수행하기 위해 토큰을 임베딩이라는 벡터로 변환한다.

        ![토큰과 임베딩](./image/02_token_embedding.png)

## Chapter 2-1 LLM 토큰화 <a class="anchor" id="chapter2-1"></a>
1. 현재 대부분의 사용자의 언어모델을 사용하는 방법은 웹 애플리케이션으로 제공되는 채팅 인터페이스를 사용하는 것이다.
    - 예시) ChatGPT, Bing Chat, Google Bard 등

2. 이러한 모델은 한 번에 모든 출력을 만들지는 않는다.
    - 모델은 텍스트를 토큰이라는 작은 단위로 나누어 처리한다.
    - 모델은 토큰을 하나씩 생성하여 출력을 만든다.

3. 토큰은 모델의 입력에도 사용된다.
    - 모델은 입력 텍스트를 토큰으로 나누어 처리한다.
    - 모델은 입력 토큰을 임베딩이라는 벡터로 변환하여 계산을 수행한다.

### Chapter 2-1-1 토크나이져가 언어 모델의 입력을 준비하는 방법 <a class="anchor" id="chapter2-1-1"></a>
1. 입력을 프롬프트를 모델에 전달하기 전에 토크나이져를 통과시켜 더 작은 단위로 쪼개야한다.
    - 아래의 그림과 같이 토크나이져가 텍스트를 단어나 부분단어로 나눈다.
    
        ![토크나이져 예시](./image/02_tokenizer_example2.png)

### Chapter 2-1-2 LLM 다운로드하고 실행하기 <a class="anchor" id="chapter2-1-2"></a>


In [2]:
# 깃허브에서 위젯 상태 오류를 피하기 위해 진행 표시줄을 나타내지 않도록 설정합니다.
from transformers.utils import logging

logging.disable_progress_bar()

In [3]:
# 트랜스포머스에서 "microsoft/Phi-3-mini-4k-instruct" 모델과 토크나이져 로드
from transformers import AutoModelForCausalLM, AutoTokenizer

# 모델과 토크나이저를 로드합니다.
model = AutoModelForCausalLM.from_pretrained(
    "microsoft/Phi-3-mini-4k-instruct",
    device_map="cuda",
    dtype="auto"
)
tokenizer = AutoTokenizer.from_pretrained("microsoft/Phi-3-mini-4k-instruct")

2026-03-23 15:16:44.493173: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2026-03-23 15:16:44.938893: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F AVX512_VNNI AVX512_BF16 AVX_VNNI FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2026-03-23 15:16:46.433944: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2026-03-23 15:16:46.436814: I external/local_xla/xla/tsl/cu

1. 프롬프트를 준비하고 토큰으로 분리한다.
    - 토큰을 모델에 전달하여 출력을 생성한다.
    - 모델에게 새로운 토큰 20개만 생성하도록 요청한다.

In [4]:
# 프롬프트 생성
#   - 사라에게 정원 가꾸기 사고에 대해 사과하는 이메일 작성
#   - <|assistant|> 토큰은 모델이 응답을 생성하기 시작하는 위치를 나타냅니다.
prompt = "Write an email apologizing to Sarah for the tragic gardening mishap. Explain how it happened.<|assistant|>"

# 입력 프롬프트를 토큰으로 변환
#   - 토크나이저는 텍스트를 모델이 이해할 수 있는 숫자 시퀀스로 변환합니다.
#   - return_tensors="pt"는 토큰이 PyTorch 텐서로 반환되도록 지정합니다.
#   - .input_ids는 토큰의 숫자 ID 시퀀스를 추출합니다.
input_ids = tokenizer(prompt, return_tensors="pt").input_ids.to("cuda")
input_ids

tensor([[14350,   385,  4876, 27746,  5281,   304, 19235,   363,   278, 25305,
           293, 16423,   292,   286,   728,   481, 29889, 12027,  7420,   920,
           372,  9559, 29889, 32001]], device='cuda:0')

In [4]:
for id in input_ids[0]:
   print(tokenizer.decode(id))

Write
an
email
apolog
izing
to
Sarah
for
the
trag
ic
garden
ing
m
ish
ap
.
Exp
lain
how
it
happened
.
<|assistant|>


In [5]:
print(tokenizer.decode(3323))
print(tokenizer.decode(622))
print(tokenizer.decode([3323, 622]))
print(tokenizer.decode(29901))

Sub
ject
Subject
:


2. 모델이 직접 텍스트 프롬프트를 입력 받지 않는다.
    - 토크나이저가 입력 프롬프트를 처리하여 input_ids라는 토큰의 시퀀스로 변환한다.

3. input_ids에는 모델이 이해할 수 있는 일련의 정수가 포함되어있다.
    - 각 정수는 어휘사전에 있는 틀정 토큰에 매핑된다.

4. 그림으로 표현하면 아래와 같이 입력이 토큰으로 나누어져 모델에 전달된다.

    ![토큰화 과정](./image/02_tokenization_process.png)

5. 토크나이져로 분리된 토큰은 아래의 주의사항이 있다.
    - 일부 토큰은 완전한 단어이다.
    - 일부 토큰은 부분단어이다.
    - 구두점 문자도 하나의 토큰이다.
    - 공백 문자는 토큰으로 취급하지 않지만 토큰의 앞이 공백인경우 "_"를 문자앞에 추가한다.

### Chapter 2-1-3 토크나이저가 텍스트를 분할하는 방법 <a class="anchor" id="chapter2-1-3"></a>
1. 모델 설계 시에 모델 작성자가 토큰호 방법을 선택한다.
    - GPT는 BPE(Byte Pair Encoding)라는 방법을 사용한다.
    - BERT는 WordPiece라는 방법을 사용한다.

2. 토큰화 방법을 선택 후 어휘사전 크기와 특수 토큰 같은 세부 사항을 결정한다.
    - 어휘사전 크기는 모델이 사용할 수 있는 고유 토큰의 수를 결정한다.
    - 특수 토큰은 모델이 입력과 출력을 구분하는 데 사용된다.
    - 예시) [CLS], [SEP], [PAD] 등

3. 토크나이저는 특정 데이터셋에서 훈련하여 해당 데이터셋을 가장 잘 표현할 수 있는 토큰화 방법을 학습한다.
    - 훈련된 토크나이저는 모델이 훈련된 데이터와 유사한 텍스트를 처리하는 데 최적화되어 있다.
    - 영어 데이터셋에서 훈련된 토크나이져와 다국어 데이터셋에서 훈련된 토크나이져는 동일한 방법과 파라미터를 사용하더라도 다른 토큰화 결과를 낼 수 있다.

4. 토크나이져는 입력 텍스트뿐 아니라 모델이 생성하는 출력 텍스트도 토큰화한다.
    - 모델이 생성하는 출력 텍스트도 토크나이져가 처리하여 토큰으로 나누어진다.
    - 모델이 생성하는 출력 텍스트는 모델이 훈련된 데이터와 유사한 텍스트를 생성하는 경향이 있다.

### Chapter 2-1-4 단어 토큰, 부분단어 토큰, 문자 토큰, 바이트 토큰 <a class="anchor" id="chapter2-1-4"></a>
1. 단어 토큰
    - 단어 토큰은 완전한 단어를 나타내는 토큰이다.
        - 예시) "hello", "world", "token" 등
    - 토크나이저가 훈련 후에 데이터셋에 새롭게 추가된 단어를 처리할 수 없다.
        - 최대한 많은 단어를 어휘사전에 포함시키려고 하지만 새로운 단어가 계속해서 등장하기 때문에 모든 단어를 포함시키는 것은 불가능하다.
    
2. 부분단어 토큰
    - 부분단어 토큰은 단어의 일부를 나타내는 토큰이다
        - 예시) "un", "##happy", "##ness" 등
    - 부분단어 토큰은 어휘사전에 없는 단어를 처리할 수 있도록 도와준다.
        - 예시) "unhappy"라는 단어가 어휘사전에 없지만 "un"과 "##happy"라는 부분단어 토큰이 어휘사전에 있다면 "unhappy"라는 단어를 "un"과 "##happy"로 나누어 처리할 수 있다.

3. 문자 토큰
    - 문자 토큰은 단어의 개별 문자를 나타내는 토큰이다
        - 예시) "h", "e", "l", "o" 등
    - 문자 토큰은 어휘사전에 없는 단어를 처리할 수 있도록 도와준다.
        - 예시) "hello"라는 단어가 어휘사전에 없지만 "h", "e", "l", "o"라는 문자 토큰이 어휘사전에 있다면 "hello"라는 단어를 "h", "e", "l", "o"로 나누어 처리할 수 있다.
    - hello를 조합하는 정보까지 모델링해야하는 부담이 있다.
    
4. 바이트 토큰
    - 유니코드 문자를 표현하는 바이트 시퀀스를 나타내는 토큰이다
        - 예시) "h"는 0x68, "e"는 0x65, "l"는 0x6C, "o"는 0x6F 등

5. 토큰화 방법을 이미지로 표현하면 아래와 같다.

    ![토큰화 방법](./image/02_tokenization_methods.png)

6. 일부 부분단어 토크나이져는 어휘사전에 바이트를 토큰으로 포함하기도 한다.
    - 표현할 수 없는 문자가 있을 때 바이트 토큰을 사용하여 처리할 수 있다.

### Chapter 2-1-5 훈련된 LLM 토크나이져 비교하기 <a class="anchor" id="chapter2-1-5"></a>

In [5]:
from transformers import AutoModelForCausalLM, AutoTokenizer

colors_list = [
    '102;194;165', '252;141;98', '141;160;203',
    '231;138;195', '166;216;84', '255;217;47'
]

def show_tokens(sentence, tokenizer_name):
    tokenizer = AutoTokenizer.from_pretrained(tokenizer_name)
    token_ids = tokenizer(sentence).input_ids
    for idx, t in enumerate(token_ids):
        # 텍스트를 인코딩한 후 다시 디코딩했을 때 원본 텍스트와 동일해지려면
        # clean_up_tokenization_spaces를 False로 지정해야 합니다.
        # 현재 이 매개변수의 기본값은 None(True에 해당)이며
        # transformers 4.45에서 True로 바뀔 예정입니다.
        # https://github.com/huggingface/transformers/issues/31884
        print(
            f'\x1b[0;30;48;2;{colors_list[idx % len(colors_list)]}m' +
            tokenizer.decode(t, clean_up_tokenization_spaces=False) +
            '\x1b[0m',
            end=' '
        )

In [7]:
text = """
English and CAPITALIZATION
🎵 鸟
show_tokens False None elif == >= else: two tabs:"		" four spaces:"    "
12.0*50=600
"""

1. BERT 베이스모델(uccased)
    - 모든문자를 소문자로 변경한다.
    - 토큰화방법: wordPice
    - 어휘사전 크기: 30,522
    - 특수토큰
        - [PAD]: 패딩 토큰, 입력 시퀀스의 길이를 맞추기 위해 사용된다.
        - [UNK]: 알 수 없는 토큰, 어휘사전에 없는 단어를 처리하기 위해 사용된다.
        - [CLS]: 분류작업을 위한 특수토큰
        - [SEP]: 문장의 구분을 나타내는 토큰
        - [MASK]: 훈련 과정 동안 일부 토큰을 감추기 위해 사용되는 마스킹 토큰
    - 줄바꿈 문자가 사라져서 모델이 채팅 같이 줄바꿈이 중요한 텍스트를 처리하는 데 어려움이 있다.
    - 'capaitalization'이 'capaital'와 '##ization'으로 나누어진다.
        - '##'문자는 이토큰이 앞에 있는 토큰의 부분토큰일을 나타낸다
        - '##'문자가 없는 토큰은 앞에 공백이 있는 토큰을 나타낸다.

In [ ]:
# BERT 베이스모델(uccased)
show_tokens(text, "bert-base-uncased")

[CLS] english and capital ##ization [UNK] [UNK] show _ token ##s false none eli ##f = = > = else : two tab ##s : " " four spaces : " " 12 . 0 * 50 = 600 [SEP] 

2. BERT 베이스모델(cased)
    - 모든 문자를 대문자로 변경한다.
    - 토큰화방법: wordPice
    - 어휘사전 크기: 28,996
    - 특수토큰은 BERT 베이스모델(uccased)와 동일하다.
    - 'CAPITALIZATION'이 CA ##PI ##TA ##L ##I ##Z ##AT ##ION으로 나누어진다.

In [8]:
# BERT 베이스모델(cased)
show_tokens(text, "bert-base-cased")

[CLS] English and CA ##PI ##TA ##L ##I ##Z ##AT ##ION [UNK] [UNK] show _ token ##s F ##als ##e None el ##if = = > = else : two ta ##bs : " " four spaces : " " 12 . 0 * 50 = 600 [SEP] 

3. GPT-2(2019)
    - 토큰화 방법: BPE(Byte Pair Encoding)
    - 어휘사전 크기: 50,257
    - 특수토큰
        - <|endoftext|>: 텍스트의 끝을 나타내는 토큰
    - 줄바꿈이 토크나이져 내에서 표현된다.
    - 대문자는 유지되고 'CAPITALIZATION'이 CAP ITAL IZ ATION 으로 나누어진다.
    - "🎵"문자는 " � � �"로 출력되었지만 각자 다른 토큰이다.
        - tokenizer.decode([8582,236,113])를 실행하면 "🎵"가 출력된다

In [9]:
# GPT-2(2019)
show_tokens(text, "gpt2")


 English  and  CAP ITAL IZ ATION 
 � � �  � � � 
 show _ t ok ens  False  None  el if  ==  >=  else :  two  tabs :" 	 	 "  four  spaces :"        " 
 12 . 0 * 50 = 600 
 

4. 코드를 이해하거나 생성하는 모델은 공백 문자가 중요하다.
    - 단일 토큰으로 네 개의 연속된 공백 문자를 표현하는 모델이 파이썬 코드 데이터셋에 더 잘맞는다.

5. Flan-T3 (2022)
    - 토큰화 방법: SentencePiece
    - 어휘사전 크기: 32,000
    - 특수토큰
        - <unk>: 알 수 없는 토큰, 어휘사전에 없는 단어를 처리하기 위해 사용된다.
        - <pad>: 패딩 토큰, 입력 시퀀스의 길이를 맞추기 위해 사용된다.
    - 줄바꿈이나 공백토큰이 없다.
    - 이모지와 한자가 모두 <unk>로 표현된다.

In [10]:
# Flan-T3 (2022)
show_tokens(text, "google/flan-t5-small")

English and CA PI TAL IZ ATION  <unk>  <unk> show _ to ken s Fal s e None  e l if = = > = else : two tab s : " " four spaces : " " 12. 0 * 50 = 600  </s> 

6. GPT-4(2023)
    - 토큰화 방법: BPE(Byte Pair Encoding)
    - 어휘사전 크기: 100,000 이상
    - 특수토큰
        - <|endoftext|>: 텍스트의 끝을 나타내는 토큰
        - <|fim_prefix|>, <|fim_middle|>, <|fim_suffix|>: 중간 토큰을 채우도록 설계된 특수 토큰
            - 세 계의 특수 토큰을 사용해 앞의 텍스트뿐만 아니라 뒤에 나오는 켁스트를 고려해 출력을 생성할 수 있다. 
    - 네 개의 공백을 하나의 토큰으로 표현한다.
    - "elseif"는 하나의 토큰으로 표현된다.
        - 자연어 외에 코드 데이터셋에서도 훈련된 모델이기 때문에 "elseif"와 같은 프로그래밍 언어의 키워드도 하나의 토큰으로 표현한다.
    - "CAPITALIZATION"이 CAPITAL IZATION 으로 나누어진다.
        - 더 적은 토큰을 사용하여 표현된다.

In [11]:
# GPT-4(2023)
show_tokens(text, "Xenova/gpt-4")


 English  and  CAPITAL IZATION 
 � � �  � � � 
 show _tokens  False  None  elif  ==  >=  else :  two  tabs :" 	 	 "  four  spaces :"      "
 12 . 0 * 50 = 600 
 

7. StarCoder2(2024)
    - 토큰화 방법: BPE(Byte Pair Encoding)
    - 어휘사전 크기: 49,152
    - 특수토큰
        - <|endoftext|>: 텍스트의 끝을 나타내는 토큰
        - <|fim_prefix|>, <|fim_middle|>, <|fim_suffix|>: 중간 토큰을 채우도록 설계된 특수 토큰
            - 세 계의 특수 토큰을 사용해 앞의 텍스트뿐만 아니라 뒤에 나오는 켁스트를 고려해 출력을 생성할 수 있다. 
    - 한 파일에서 다른 파일에 정의된 함수를 호출할 수 있다.
        - 모델은 같은 장소의 다른 파일에 있는 코드를 식별하고 다른 저장소에 있는 코드와 구반할 수 있어야한다.
    - 코드 생성에 초점을 맞추 디코더
    - 여러개의 공백을 하나의 토큰으로 표현한다.

In [12]:
# StarCoder2(2024)
show_tokens(text, "bigcode/starcoder2-15b")


 English  and  CAPITAL IZATION 
 � � �   � � 
 show _ tokens  False  None  elif  ==  >=  else :  two  tabs :" 	 	 "  four  spaces :"      " 
 1 2 . 0 * 5 0 = 6 0 0 
 

8. Galactica
    - 토큰화 방법: BPE(Byte Pair Encoding)
    - 어휘사전 크기: 50,000
    - 특수토큰
        - <s> : 텍스트의 시작을 나타내는 토큰
        - </s> : 텍스트의 끝을 나타내는 토큰
        - <pad> : 패딩 토큰, 입력 시퀀스의 길이를 맞추기 위해 사용된다.
        - <unk> : 알 수 없는 토큰, 어휘사전에 없는 단어를 처리하기 위해 사용된다.


In [13]:
# Galactica
show_tokens(text, "facebook/galactica-1.3b")


 English  and  CAP ITAL IZATION 
 � � � �  � � � 
 show _ tokens  False  None  elif   ==   > =  else :  two  t abs : " 		 "  four  spaces : "      " 
 1 2 . 0 * 5 0 = 6 0 0 
 

9. Phi-3 (Llama 2)
    - 토큰화 방법: BPE(Byte Pair Encoding)
    - 어휘사전 크기: 32,000
    - 특수토큰
        - <|endoftext|>: 텍스트의 끝을 나타내는 토큰
        - 채팅토큰: 대화 차례와 역할을 나타내는 특수 토큰
            - 예시) <|system|>, <|user|>, <|assistant|> 등

In [14]:
# Phi-3 (Llama 2)
show_tokens(text, "microsoft/Phi-3-mini-4k-instruct")

 
 English and C AP IT AL IZ ATION 
 � � � �  � � � 
 show _ to kens False None elif == >= else : two tabs :" 	 	 " four spaces :"    " 
 1 2 . 0 * 5 0 = 6 0 0 
 

### Chapter 2-1-6 토크나이저 속성 <a class="anchor" id="chapter2-1-6"></a>


## Chapter 2-2 토큰 임베딩 <a class="anchor" id="chapter2-2"></a>
1. 언어는 토큰의 시퀸스로 볼 수 있다.
    - 좋은 모델을 충분히 큰 토큰 집합에서 훈련한다면 모델이 언어의 복잡한 패턴을 학습할 수 있다.

2. 다음 단계는 토큰을 위한 최상의 수치표현을 찾는 것 이다.
    - 모델은 이를 사용해 텍스트에 있는 패턴을 계산하고 적절히 모델링 할 수 있다.
    - 이런 패턴은 특정 언어에 대한 모델의 일관성, 코드 생상 능력 등으로 들어난다.

3. 수치표현을 임베딩이라고한다.
    - 임베딩은 언어에 있는 의미와 패턴을 포착하기 위한 수치 표현공간이다.

### Chapter 2-2-1 토크나이저의 어휘사전에 대한 임베딩을 내장한 언어 모델 <a class="anchor" id="chapter2-2-1"></a>
1. 토크나이저가 초기화되고 훈련되고 나면 이를 사용해 언어 모델을 훈련한다.
    - 사전 훈련된 언어 모델이 해당 토크나이저와 연결되는 이유이다.
    - 모델을 재훈련하지 않고는 다른 토크나이저를 사용할 수 없다.

2. 언어 모델은 토크나이저의 어휘사전에 있는 각 토큰에 대한 임베딩 벡터를 가지고 있다.
    - 사전 훈련된 모델 중 일부는 일런 벡터가 모두 담긴 임베딩 행렬을 포함한다.
    
        ![임베딩 행렬](./image/02_embedding_matrix.png)

3. 훈련 과정이 시작되기 전에 이 벡터들은 랜덤하게 초기화된다.
    - 모델이 훈련되면서 이 벡터들은 언어의 패턴과 의미를 포착하도록 조정된다.
    - 모델이 훈련된 후에는 토큰의 임베딩 벡터가 언어의 의미와 패턴을 포착하는 수치 표현이 된다.

### Chapter 2-2-2 언어 모델로 문맥을 고려한 단어 임베딩 만들기 <a class="anchor" id="chapter2-2-2"></a>
1. 토큰 임베딩을 생성하는 것은 언어 모델을 텍스트 표현에 사용하는 주요 방법 중 하나이다.
    - 이를 통해 개체명 인식이나 추출적 텍스트 요약같은 작업이 가능하다.
    - 추출적 텍스트: 모델이 입력 텍스트에서 중요한 정보를 추출하여 요약하는 작업

2. 각 토큰이나 단어를 정적인 벡터로 나타내는 대신 언어 모델은 문맥을 고려한 토큰 임베딩을 만든다.
    - 단어는 문맥에 따라 다른 임베딩으로 표현된다.
    
    ![문맥을 고려한 토큰 임베딩](./image/02_contextual_token_embedding.png)

In [1]:
# 깃허브에서 위젯 상태 오류를 피하기 위해 진행 표시줄을 나타내지 않도록 설정합니다.
from transformers.utils import logging

logging.disable_progress_bar()

/home/redinblue/anaconda3/envs/python312/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [2]:
from transformers import AutoModel, AutoTokenizer

# 언어 모델을 로드합니다.
model = AutoModel.from_pretrained("microsoft/deberta-v3-xsmall")

DebertaV2Model LOAD REPORT from: microsoft/deberta-v3-xsmall
Key                                        | Status     |  | 
-------------------------------------------+------------+--+-
lm_predictions.lm_head.LayerNorm.bias      | UNEXPECTED |  | 
mask_predictions.LayerNorm.bias            | UNEXPECTED |  | 
mask_predictions.classifier.weight         | UNEXPECTED |  | 
lm_predictions.lm_head.LayerNorm.weight    | UNEXPECTED |  | 
mask_predictions.LayerNorm.weight          | UNEXPECTED |  | 
deberta.embeddings.word_embeddings._weight | UNEXPECTED |  | 
mask_predictions.dense.bias                | UNEXPECTED |  | 
lm_predictions.lm_head.dense.weight        | UNEXPECTED |  | 
lm_predictions.lm_head.bias                | UNEXPECTED |  | 
lm_predictions.lm_head.dense.bias          | UNEXPECTED |  | 
mask_predictions.classifier.bias           | UNEXPECTED |  | 
mask_predictions.dense.weight              | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/a

In [3]:
!pip list | grep transformers

sentence-transformers   5.3.0
transformers            5.0.0


In [ ]:
# 토크나이저를 로드합니다.
tokenizer = AutoTokenizer.from_pretrained("microsoft/deberta-v3-base")

# 문장을 토큰으로 나눈다.
tokens = tokenizer("Hello world", return_tensors="pt")

# 토큰을 처리한다.
#   - **tokens는 토큰 딕셔너리를 언패킹하여 모델에 전달하는 것을 의미합니다.
outputs = model(**tokens)[0]

In [22]:
outputs.shape

torch.Size([1, 2, 384])

3. "deberta-v3-base" 모델은  토큰 임베딩을 만드는 최고의 언어모델 중 하나이다.
    - "Hello world" 문장의 출력 shpape는 (1, 2, 384)이다.
    - 첫 번째 차원은 배치 크기이며, 두 번째 차원은 두 개의 토큰이 있고, 각 토큰은 384차원의 벡터로 임베딩되었다.

In [12]:
for token in tokens['input_ids'][0]:
    print(tokenizer.decode(token, skip_special_tokens=True))

Hello
world


4. 기술적으로 언어 모델 안에서 일어나는 첫 번째 단계는 토큰 ID를 원시 임베딩으로 변환하는 것이다.
    - 언어 모델은 정적인 원시 임베딩을 사용해 문맥을 고려한 텍스트 임베딩을 만든다.
    
        ![토큰 ID를 원시 임베딩으로 변환하는 과정](./image/02_token_id_to_embedding.png)

## Chapter 2-3 텍스트 임베딩 <a class="anchor" id="chapter2-3"></a>
1. LLM 애플리케이션은 전체 문장, 단어, 문단 또는 텍스트 문서를 나타내는 텍스트 임베딩을 필요로 하는 경우가 많다.
    - 예시) 문서 검색, 텍스트 분류, 감정 분석 등

2. 텍스트 임베딩 모델은 텍스트 조각을 입력받아 텍스트를 표현하고 유용한 의미를 포착하여 벡터로 변환하는 모델이다.
    - 임베딩 모델을 사용해 특성을 추출하고 입력 텍스트를 임베딩으로 변환한다.

        ![텍스트 임베딩](./image/02_text_embedding.png)

3. 텍스트 임베딩 벡터를 만드는 방법에는 여러가지가 있다.

4. sentence-transformers 라이브러리를 사용해 텍스트 임베딩을 만들 수 있다.
    - 사전 훈련된 임베딩 모델을 활용하는 데 인기가 많은 페키지 이다.
    - 공개된 모델을 로드할 수 있다.

In [2]:
# all-mpnet-base-v2 모델을 sentence-transformers 라이브러리를 사용해 로드합니다.
from sentence_transformers import SentenceTransformer

#  모델을 로드한다.
model = SentenceTransformer('sentence-transformers/all-mpnet-base-v2')

# 텍스트를 텍스트 임베딩으로 변환한다.
vector = model.encode("Best movie ever!")

Loading weights: 100%|██████████| 199/199 [00:00<00:00, 540.02it/s, Materializing param=pooler.dense.weight]                        
MPNetModel LOAD REPORT from: sentence-transformers/all-mpnet-base-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


In [3]:
vector.shape

(768,)

5. "Best movie ever!"라는 텍스트가 768차원의 벡터로 표현된다.

## Chapter 2-4 LLM을 넘어 활용되는 단어 임베딩 <a class="anchor" id="chapter2-4"></a>
1. 임베딩은 텍스트와 언어 생성 등의 작업 외에도 유용하다.
    - 임베딩 또는 객체에 의미 있는 벡터 표현을할당하는 것은 추천 엔진과 로봇공학을 포함해 많은 분야에서 유용하다.

2. 사전 훈련된 word2vec 임베딩을 사용하는 방법과 이 알고리즘이 단어 임베딩을 어떻게 생성하는지 알아보자


### Chapter 2-4-1 사전 훈련된 단어 임베딩 다운로드 <a class="anchor" id="chapter2-4-1"></a>
1. Gensim 라이브러리를 사용해 시전훈련된 단어 임베딩을 다운로드하는  방법을 알아본다.

In [5]:
import gensim.downloader as api

#  위키백과에서 훈련된 GloVe 임베딩을 로드합니다.
model = api.load("glove-wiki-gigaword-50")

[==================================================] 100.0% 66.0/66.0MB downloaded


In [6]:
# king 과 같은 특정 단어와 가장 가까운 이웃 단어를 살펴본다.
model.most_similar([model['king']], topn=11)

[('king', 1.0000001192092896),
 ('prince', 0.8236179351806641),
 ('queen', 0.7839043140411377),
 ('ii', 0.7746230363845825),
 ('emperor', 0.7736247777938843),
 ('son', 0.766719400882721),
 ('uncle', 0.7627150416374207),
 ('kingdom', 0.7542161345481873),
 ('throne', 0.7539914846420288),
 ('brother', 0.7492411136627197),
 ('ruler', 0.7434253692626953)]

### Chapter 2-4-2 word2vec 알고리즘과 대조 훈련 <a class="anchor" id="chapter2-4-2"></a>
1. LLM과 비슷하게 word2vec은 텍스느로부터 만든 샘플에서 훈련된다.
    - 예를 들어 "Thos shalt not make a machine in the likeness of a human mind"라는 문장이 있다면.
    - 슬라이딩 윈도를 사용해 훈련 샘플을 생성한다.
    - 윈도우 크기가 2이면 중심 단어의 앞뒤로 이어질 두 개의 단어를 고려한다.

2. 임베딩은 분류 작을 통해 생성된다.
    - 단어가 일반적으로 동일한 문맥에서 나타나는지 예측하도록 신경망을 훈련한다.
        - 문맥은 모델링하려고 훈련 데이터넷에 있는 많은 문장의 문맥을 의미한다.
    - 신경망이 두 개의 단어를 입력받고 두 단어가 동일한 문맥에 등장하면 1, 그렇지 않으면 0을 출력한다고 가정한다.

3. 아래의 그림과 같이 첫 번째 슬라이딩 윈도우에서 네 개의 훈련 샘풀을 생성할 수 있다.
    - 슬라이딩 윈도를 사용해 훈련 샘플을 만들고 word2vec 알고리즘은 두 단어가 이웃인지 아닌지 예측한다.

        ![슬라이딩 윈도우를 사용해 훈련 샘플 생성](./image/02_sliding_window_training_samples.png)

4. 생성된 각각의 훈련 샘플은 중심에 있는 단어를 하나의 입력으로 사용하고, 이웃 단어를 두 번재 입력으로 사용한다.
    - 최종 훈련된 모델은 이웃 관계를 감지하여, 두 입력 단어가 이웃이면 1을 출력할 것이 기대한다.
    - 이웃 단어는 일반적으로 동일한 문맥에서 나타나는 단어를 의미한다.

        ![입력 샘플](./image/02_input_samples.png)


5. 데이터셋에 타깃 값이 1만 있다면 모델이 속임수를 배워 항상 1만 출력할 수 있다.
    - 이웃이 아닌 단어로 만든 샘플을 사용해 훈련 데이터셋을 보안해야 한다.
    - 음성 샘플을 추가한 훈련 데이터셋을 사용해 모델이 이웃이 아닌 단어를 구별하도록 훈련한다.

        ![음성 샘플 추가](./image/02_negative_samples.png)

6. 음성 샘플을 선택할 때 굳이 과학정일 필요는 없다고 알려져있다.
    - 랜덤하게 생성된 샘플에서 양성 샘플을 감지하는 간단한 능력을 바탕으로 유용한 모젱델을 만들 수 있다.

7. word2vec의 중요한 두 가지 개념을 정리하였다.
    - 이웃 단어를 선택하는 방법인 스킵그램
    - 데이터셋에서 랜덤하게 샘플링하여 음성 샘플을 만드는 네거티브 샘플링

        ![word2vec의 스킵그램과 네거티브 샘플링](./image/02_word2vec_skipgram_negative_sampling.png)

8. 텍스트로 부터 이와 같은 샘플을 수백만 내지 수십억 개 만들 수 있다.
    - 이 데이터셋으로 신경망을 훈련하기 전에 LLM 토크나이정서 보았던 것 처럼 몇 가지 토큰화에 대한 결정을 내려야한다.
    - 아래 그림과 같이 토큰을 위한 임베딩 벡터를 랜덤하게 초기화 하여야 한다.
    
        ![토큰 임베딩 벡터 랜덤 초기화](./image/02_token_embedding_random_initialization2.png)

9. 두 개의 임베딩 벡터를 사용해 이웃한 단어인지 아닌지를 예측하도록 모델을 훈련한다.
    - 예측이 맞는지 틀리는지에 따라 전형적인 머신러닝 훈련 단계가 임베딩을 업데이트하여 다음 번에 모델이 두 벡터를 만났을 때 올바른 답을 낼 가능성을 높인다.
    - 훈련 과장이 끝나면 어휘 사전에 있는 모든 토큰의 임베딩이 향상된다.

        ![이웃한 단어 예측](./image/02_predicting_neighboring_words.png)

## Chapter 2-5 추천시스템을 위한 시스템 임베딩 <a class="anchor" id="chapter2-5"></a>
### Chapter 2-5-1 임베딩으로 노래 추천하기 <a class="anchor" id="chapter2-5-1"></a>
1. word2vec 알고리즘을 사용해 사람들이 만든 음악 재생목록에 있는 노래를 임베딩한다.
    - 각각의 노래를 단어나 토큰으로 다루고 재생목록을 문장처럼 취급한다고 생각해보자.
    - 그 다음 이 임베딩을 사용해 재생목록에 자주 함께 등장하는 비슷한노래를 추천할 수 있다.

2. 코넬 대학교의 유오 첸이 모은 데이터넷을 사용한다.
    - 미국 전영게 있는 수백 개의 라디오 방속국에서 가져온 재생목록이 담겨져있다.
    - 노래의 유사도를 포착하는 임베딩을 위해 재생목록을 모아 구성한 데이터셋을 사용한다.
    - 각 재생목록에는 노래 목록이 담겨져 있다

        ![음악 재생목록 데이터셋](./image/02_music_playlist_dataset.png)

3. 구축 방법을 살펴보기 전에 최종 제품을 시연해 본다.
    - 마이클 잭슨의 "Billie Jean"이라는 노래를 입력으로 사용해 보자. 
    - ID는 3822번이다.
    - 마돈나, 플린스, 마이클 책슨의 노래 몇곡이 가장 가까운 이웃이다.
    
        ![노래 추천 결과](./image/02_song_recommendation_results.png)

### Chapter 2-5-2 노래 임베딩 모델 훈련하기 <a class="anchor" id="chapter2-5-2"></a>
1. 먼저 제목과 아티스트 같은 노래의 메타데이터와 노래의 재생목록을 담고 있는 데이터셋을 로드한다.

In [39]:
import pandas as pd
from urllib import request

# 재생목록  데이터셋 파일을 가져온다.
data = request.urlopen('https://storage.googleapis.com/maps-premium/dataset/yes_complete/train.txt')

# 재생목록 파일을 파싱한다. 처음 두 줄은 메타데이터만 담고 있으므로 건너뛴다.
lines = data.read().decode("utf-8").split('\n')[2:]

# 하나의 노래만 있는 재생목록은 삭제한다.
#    - rstrip()은 각 줄의 오른쪽 끝에서 공백 문자를 제거합니다.
playlists = [ s.rstrip().split() for s in lines if len(s.split()) > 1]
print(playlists[:5])

# 노래의 메타데이터를 로드한다.
songs_file = request.urlopen('https://storage.googleapis.com/maps-premium/dataset/yes_complete/song_hash.txt')
songs_file = songs_file.read().decode("utf-8").split('\n')

# 노래의 ID, 제목, 아티스트를 로드한다.
songs = [s.rstrip().split('\t') for s in songs_file]
print(songs[:5])

songs_df = pd.DataFrame(data=songs, columns=['id', 'title', 'artist']).set_index('id')

# 노래의 ID를 인덱스로 설정한다.
songs_df.head()

[['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31', '32', '33', '34', '35', '36', '37', '38', '39', '40', '41', '2', '42', '43', '44', '45', '46', '47', '48', '20', '49', '8', '50', '51', '52', '53', '54', '55', '56', '57', '25', '58', '59', '60', '61', '62', '3', '63', '64', '65', '66', '46', '47', '67', '2', '48', '68', '69', '70', '57', '50', '71', '72', '53', '73', '25', '74', '59', '20', '46', '75', '76', '77', '59', '20', '43'], ['78', '79', '80', '3', '62', '81', '14', '82', '48', '83', '84', '17', '85', '86', '87', '88', '74', '89', '90', '91', '4', '73', '62', '92', '17', '53', '59', '93', '94', '51', '50', '27', '95', '48', '96', '97', '98', '99', '100', '57', '101', '102', '25', '103', '3', '104', '105', '106', '107', '47', '108', '109', '110', '111', '112', '113', '25', '63', '62', '114', '115', '84', '116', '117', '118', '119', '120', '121', '

,title,artist
id,,
0,Gucci Time (w\/ Swizz Beatz),Gucci Mane
1,Aston Martin Music (w\/ Drake & Chrisette Mich...,Rick Ross
2,Get Back Up (w\/ Chris Brown),T.I.
3,Hot Toddy (w\/ Jay-Z & Ester Dean),Usher
4,Whip My Hair,Willow


In [13]:
# playlists는 각 재생목록이 노래 ID의 시퀀스로 표현된 리스트입니다.
print('재생목록 #1: ', playlists[0], '\n')
print('재생목록 #2: ', playlists[1],)

재생목록 #1:  ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31', '32', '33', '34', '35', '36', '37', '38', '39', '40', '41', '2', '42', '43', '44', '45', '46', '47', '48', '20', '49', '8', '50', '51', '52', '53', '54', '55', '56', '57', '25', '58', '59', '60', '61', '62', '3', '63', '64', '65', '66', '46', '47', '67', '2', '48', '68', '69', '70', '57', '50', '71', '72', '53', '73', '25', '74', '59', '20', '46', '75', '76', '77', '59', '20', '43'] 

재생목록 #2:  ['78', '79', '80', '3', '62', '81', '14', '82', '48', '83', '84', '17', '85', '86', '87', '88', '74', '89', '90', '91', '4', '73', '62', '92', '17', '53', '59', '93', '94', '51', '50', '27', '95', '48', '96', '97', '98', '99', '100', '57', '101', '102', '25', '103', '3', '104', '105', '106', '107', '47', '108', '109', '110', '111', '112', '113', '25', '63', '62', '114', '115', '84', '116', '117', '118', '1

2. playlists에는 여러 사용자의 재송목록이 노래 ID의 시퀀스로 표현된 리스트가 담겨져 있다.
    - word2vec 알로리즘으로 훈련하면 특정 노래의 ID를 입력하여 비슷한 노래의 ID를 출력할 수 있다.
    - 이 ID를 사용해 노래의 제목과 아티스트를 찾을 수 있다.

In [21]:
# word2vec을 훈련한다.
from gensim.models import Word2Vec

# Word2Vec 모델을 훈련한다.
#   - playlists는 훈련 데이터로 사용됩니다. 각 재생목록은 노래 ID의 시퀀스로 표현됩니다.
#   - vector_size는 각 단어(노래 ID)가 임베딩될 벡터의 차원 수를 지정합니다.
#   - window는 컨텍스트 윈도우의 크기를 지정합니다. 이 경우, 모델은 각 노래 ID 주변의 최대 20개의 노래 ID를 컨텍스트로 사용합니다.
#   - negative는 네거티브 샘플링의 수를 지정합니다. 이 경우, 모델은 각 긍정적 훈련 샘플에 대해 50개의 부정적 샘플을 사용합니다.
#   - min_count는 단어가 훈련에 포함되기 위한 최소 빈도를 지정합니다. 이 경우, 모든 단어가 포함됩니다.
#   - workers는 훈련에 사용할 CPU 코어 수를 지정합니다. 이 경우, 4개의 코어가 사용됩니다.
model = Word2Vec(sentences=playlists, vector_size=32, window=20, negative=50, min_count=1, workers=4)


In [43]:
# id=2172와 비슷한 노래 ID를 찾는다.
song_id = '2172'

model.wv.most_similar(positive=str(song_id))

[('3167', 0.9971422553062439),
 ('2849', 0.9964052438735962),
 ('3119', 0.9958793520927429),
 ('2976', 0.995421826839447),
 ('5586', 0.995011568069458),
 ('5634', 0.9947982430458069),
 ('1922', 0.994246780872345),
 ('2640', 0.9940305352210999),
 ('10084', 0.9936514496803284),
 ('2704', 0.9933838248252869)]

In [40]:
# 2172가 어떤 노래인지 확인한다.
print(songs_df.iloc[2172])

title     Fade To Black
artist        Metallica
Name: 2172 , dtype: str


In [ ]:
# 추천노래 출력 함수 만들기
import numpy as np

def print_recommendations(song_id):
    similar_songs = np.array(model.wv.most_similar(positive=str(song_id), topn=5))[:,0]
    print(f"노래 ID {song_id}와 비슷한 노래 ID: {similar_songs}")
    return songs_df.iloc[similar_songs]

print_recommendations(2172)


노래 ID 2172와 비슷한 노래 ID: ['3167' '2849' '3119' '2976' '5586']


,title,artist
id,,
3167,Unchained,Van Halen
2849,Run To The Hills,Iron Maiden
3119,There's Only One Way To Rock,Sammy Hagar
2976,I Don't Know,Ozzy Osbourne
5586,The Last In Line,Dio
